File: cpp/_327.cpp
class FenwickTree {    
public:
    FenwickTree(int n): sums_(n + 1, 0) {}
    
    void update(int i, int delta) {
        while (i < sums_.size()) {
            sums_[i] += delta;
            i += lowbit(i);
        }
    }
    
    int query(int i) const {        
        int sum = 0;
        while (i > 0) {
            sum += sums_[i];
            i -= lowbit(i);
        }
        return sum;
    }
private:
    static inline int lowbit(int x) { return x & (-x); }
    vector<int> sums_;
};

// Solution 1. Binary Index Tree (Fenwick Tree)
// O(NlogN) / O(N)
class Solution {
public:
    int countRangeSum(vector<int>& nums, int lower, int upper) {
        int n = nums.size();
        
        // construct presum
        vector<long long> p(n + 1, 0);
        for (int i = 0; i < n; ++i)
            p[i + 1] = p[i] + nums[i];
        
        // sort presum array for binary searching
        sort(p.begin(), p.end());
        
        // BIT stores the index of sorted presum (from 1)
        FenwickTree fw_tree(n + 1);
        
        // sum up the number of valid presums for each presum
        int ans = 0;
        long long cur = 0;
        fw_tree.update(lower_bound(p.begin(), p.end(), 0) - p.begin() + 1, 1);
        for (int i = 0; i < n; ++i) {
            cur += nums[i];
            int left = lower_bound(p.begin(), p.end(), cur - upper) - p.begin();
            int right = upper_bound(p.begin(), p.end(), cur - lower) - p.begin();
            ans += (fw_tree.query(right) - fw_tree.query(left));
            fw_tree.update(lower_bound(p.begin(), p.end(), cur) - p.begin() + 1, 1);
        }
        
        return ans;
    }
};


// Solution 2. Divide and Conquer
// O(NlogN) / O(N)
class Solution {
public:
    int countRangeSum(vector<int>& nums, int lower, int upper) {
        int n = nums.size();
        vector<long> p(n + 1, 0);
        for(int i =0; i < n; ++i)
            p[i + 1] = p[i] + nums[i];
        return mergeSort(p, lower, upper, 0, n + 1);
    }
private:
    int mergeSort(vector<long>& p, int lower, int upper, int low, int high) {
        if (low + 1 >= high) return 0;

        int mid = low + (high - low) / 2;
        int ans = mergeSort(p, lower, upper, low, mid) + 
            mergeSort(p, lower, upper, mid, high);
        int l = mid, r = mid;

        for (int i = low; i < mid; ++i) {
            while (l < high && p[l] - p[i] < lower) l++;
            while (r < high && p[r] - p[i] <= upper) r++;
            ans += r - l;
        }
        inplace_merge(p.begin() + low, p.begin() + mid, p.begin() + high);
        return ans;
    }
};

File: python/_327.py
class FenwickTree:
    def __init__(self, n):
        self.sums = [0] * (n + 1)
        
    def update(self, i, delta):
        while i < len(self.sums):
            self.sums[i] += delta
            i += self._lowbit(i)
    
    def query(self, i):
        s = 0
        while i > 0:
            s += self.sums[i]
            i -= self._lowbit(i)
        return s
        
    def _lowbit(self, x):
        return x & -x

# Solution 1. Binary Index Tree (Fenwick Tree)
# O(NlogN) / O(N)
class Solution:
    def countRangeSum(self, nums: List[int], lower: int, upper: int) -> int:
        n = len(nums)
        
        # construct presum
        presum = [0] * (n + 1)
        for i in range(n):
            presum[i + 1] = presum[i] + nums[i]
            
        # sort presum array for binary searching
        presum.sort()
        
        # BIT stores the index of sorted presum (from 1)
        fw_tree = FenwickTree(n + 1)
        
        # sum up the number of valid presums for each presum
        ans = cur = 0
        fw_tree.update(bisect.bisect_left(presum, 0) + 1, 1)
        for i in range(n):
            cur += nums[i]
            left = fw_tree.query(bisect.bisect_left(presum, cur - upper))
            right = fw_tree.query(bisect.bisect_right(presum, cur - lower))
            ans += (right - left)
            fw_tree.update(bisect.bisect_left(presum, cur) + 1, 1)
        
        return ans


# Solution 2. Divide and Conquer
# O(NlogN) / O(N)
class Solution:
    def countRangeSum(self, nums: List[int], lower: int, upper: int) -> int:
        n = len(nums)
        presum = [0] * (n + 1)
        for i in range(n):
            presum[i + 1] = presum[i] + nums[i]
            
        def merge_sort(low, high):
            if low + 1 >= high: return 0
            
            mid = low + (high - low) // 2
            ans = merge_sort(low, mid) + merge_sort(mid, high)
            l = r = mid
            
            for i in range(low, mid):
                while l < high and presum[l] - presum[i] < lower:
                    l += 1
                while r < high and presum[r] - presum[i] <= upper:
                    r += 1
                ans += r - l
            presum[low: high] = heapq.merge(presum[low: mid], presum[mid: high])
            return ans
            
        return merge_sort(0, n + 1)